### PR TITLE
COCOS - 209 - Restructure agent and manager gRPC config

### LINKED ISSUE #209: Feature: Restructure Configuration for Manager and Agent Services
## Background
We currently have two configuration structures for our manager and agent services. Both services share the same gRPC server and client library code, but attested TLS should only be usable for the agent service.

### Client Config
```go
type Config struct {
	ClientCert   string        `env:"CLIENT_CERT"     envDefault:""`
	ClientKey    string        `env:"CLIENT_KEY"      envDefault:""`
	ServerCAFile string        `env:"SERVER_CA_CERTS" envDefault:""`
	URL          string        `env:"URL"             envDefault:"localhost:7001"`
	Timeout      time.Duration `env:"TIMEOUT"         envDefault:"60s"`
	AttestedTLS  bool          `env:"ATTESTED_TLS"    envDefault:"false"`
	Manifest     string        `env:"MANIFEST"        envDefault:""`
}
```

### Server Config
```go
type Config struct {
	Host         string `env:"HOST"               envDefault:""`
	Port         string `env:"PORT"               envDefault:""`
	CertFile     string `env:"SERVER_CERT"        envDefault:""`
	KeyFile      string `env:"SERVER_KEY"         envDefault:""`
	ServerCAFile string `env:"SERVER_CA_CERTS"    envDefault:""`
	ClientCAFile string `env:"CLIENT_CA_CERTS"    envDefault:""`
	AttestedTLS  bool   `env:"ATTESTED_TLS"       envDefault:"false"`
}
```

## Proposed Changes

1. Create a base configuration structure for common fields.
2. Create separate structures for manager-specific and agent-specific configurations.
3. Embed the base configuration in both manager and agent configurations.
4. Move TLS-related fields to the agent configuration only.

## Benefits
- Clear separation of concerns between manager and agent configurations.
- attested TLS-related fields are explicitly only available for the agent service.
- Shared fields are centralized in the BaseConfig, reducing duplication.
- Easier to maintain and extend in the future.

## Tasks
- [ ] Implement the new configuration structures
- [ ] Update the manager service to use ManagerConfig
- [ ] Update the agent service to use AgentConfig
- [ ] Adjust any code that relies on the old configuration structures
- [ ] Update documentation to reflect the new configuration structure
- [ ] Add tests to ensure proper functioning with the new configuration

### Indicate the importance of this feature to you.

Must-have

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by WashingtonKK**
# What type of PR is this?

This is a refactor of agent and manager grpc configs.

## What does this do?

- Introduced a new configuration structure for gRPC clients and servers, enhancing integration and management of settings.

## Which issue(s) does this PR fix/relate to?

- Related Issue #
- Resolves #209 

## Have you included tests for your changes?

Yes

### CODE REVIEW

File: .vscode/launch.json
Comment (SammyOina) [inline .vscode/launch.json:1]: should not be tracked

File: cocos-manager.env
Comment (SammyOina) [inline cocos-manager.env:1]: these changes are not needed

File: internal/server/server.go [hunk 1]
```diff
@@ -17,9 +19,8 @@ type Server interface {
...
 }
 
-type Config struct {
-	Host         string `env:"HOST"               envDefault:""`
-	Port         string `env:"PORT"               envDefault:""`
+type AgentConfig struct {
+	grpc.BaseConfig
 	CertFile     string `env:"SERVER_CERT"        envDefault:""`
 	KeyFile      string `env:"SERVER_KEY"         envDefault:""`
 	ServerCAFile string `env:"SERVER_CA_CERTS"    envDefault:""`
```
Comment (SammyOina) [inline internal/server/server.go:26]: should be part of server base config

File: pkg/clients/grpc/connect.go [hunk 2]
```diff
@@ -63,14 +63,19 @@ var (
...
+
+type ManagerConfig struct {
+	BaseConfig
+	ClientCert   string `env:"CLIENT_CERT"     envDefault:""`
+	ClientKey    string `env:"CLIENT_KEY"      envDefault:""`
+	ServerCAFile string `env:"SERVER_CA_CERTS" envDefault:""`
+	BackendInfo  string `env:"BACKEND_INFO"    envDefault:""`
+	ClientTLS  bool   `env:"CLIENT_TLS"    envDefault:"false"`
+
 }
```
Comment (SammyOina) [inline pkg/clients/grpc/connect.go:79]: you need cli config as well(agent client)

File: pkg/clients/grpc/connect.go [hunk 3]
```diff
@@ -91,13 +96,13 @@ type Client interface {
...
 	*grpc.ClientConn
-	cfg    Config
+	cfg    ManagerConfig
 	secure security
 }
 
 var _ Client = (*client)(nil)
 
-func NewClient(cfg Config) (Client, error) {
+func NewClient(cfg ManagerConfig) (Client, error) {
```
Comment (SammyOina) [inline pkg/clients/grpc/connect.go:105]: should handle both manager and agent client configs

**Review by SammyOina — CHANGES_REQUESTED** [review]

**Comment by drasko** [comment]
Fix the title

File: internal/server/grpc/grpc.go [hunk 4]
```diff
@@ -60,8 +60,8 @@ type serviceRegister func(srv *grpc.Server)
 
 var _ server.Server = (*Server)(nil)
 
-func New(ctx context.Context, cancel context.CancelFunc, name string, config server.Config, registerService serviceRegister, logger *slog.Logger, qp client.QuoteProvider, authSvc auth.Authenticator) server.Server {
-	listenFullAddress := fmt.Sprintf("%s:%s", config.Host, config.Port)
+func New(ctx context.Context, cancel context.CancelFunc, name string, config server.AgentConfig, registerService serviceRegister, logger *slog.Logger, qp client.QuoteProvider, authSvc auth.Authenticator) server.Server {
```
Comment (SammyOina) [inline internal/server/grpc/grpc.go:63]: this function is used by computations server which is different from agent, use interface type and type assertion things specific to agent server

Comment (SammyOina) [inline internal/server/grpc/grpc.go:63]: this function is used by computations server which is different from agent, use interface type and type assertion things specific to agent server

Comment (SammyOina) [reply internal/server/grpc/grpc.go:63]: you can create a custom interface with a method such as getBaseConfig, and then do type assertion for specific fields like attested tls

Comment (SammyOina) [reply internal/server/grpc/grpc.go:63]: you can create a custom interface with a method such as getBaseConfig, and then do type assertion for specific fields like attested tls

File: pkg/clients/grpc/agent/agent.go [hunk 5]
```diff
@@ -7,15 +7,20 @@ import (
...
 	"github.com/ultravioletrs/cocos/pkg/clients/grpc"
 	grpchealth "google.golang.org/grpc/health/grpc_health_v1"
 )
 
 var ErrAgentServiceUnavailable = errors.New("agent service is unavailable")
 
 // NewAgentClient creates new agent gRPC client instance.
-func NewAgentClient(ctx context.Context, cfg grpc.Config) (grpc.Client, agent.AgentServiceClient, error) {
-	client, err := grpc.NewClient(cfg)
+func NewAgentClient(ctx context.Context, cfg server.AgentConfig) (grpc.Client, agent.AgentServiceClient, error) {
```
Comment (SammyOina) [inline pkg/clients/grpc/agent/agent.go:18]: client and server config are distinct

Comment (SammyOina) [inline pkg/clients/grpc/agent/agent.go:18]: client and server config are distinct

Comment (SammyOina) [inline pkg/clients/grpc/agent/agent.go:18]: client and server config are distinct

Comment (SammyOina) [inline pkg/clients/grpc/agent/agent.go:18]: client and server config are distinct

File: pkg/clients/grpc/manager/manager.go [hunk 6]
```diff
@@ -7,9 +7,20 @@ import (
 	"github.com/ultravioletrs/cocos/pkg/clients/grpc"
 )
 
+type ManagerConfig struct {
+	grpc.BaseConfig
+	BackendInfo string `env:"BACKEND_INFO"    envDefault:""`
```
Comment (SammyOina) [inline pkg/clients/grpc/manager/manager.go:12]: manager client config is not configured using backendinfo, this exclusive to agent client(CLI). also clienttls should be part of base client config

File: pkg/clients/grpc/connect.go [hunk 7]
```diff
@@ -49,14 +49,22 @@ var (
...
+	CertFile     string        `env:"SERVER_CERT"        envDefault:""`
+	KeyFile      string        `env:"SERVER_KEY"         envDefault:""`
+	ClientCAFile string        `env:"CLIENT_CA_CERTS"    envDefault:""`
+}
+
+type ClientConfig struct {
+	BaseConfig
+	BackendInfo string `env:"BACKEND_INFO"    envDefault:""`
+	ClientTLS   bool   `env:"CLIENT_TLS"    envDefault:"false"`
+	AttestedTLS bool   `env:"ATTESTED_TLS"       envDefault:"false"`
```
Comment (SammyOina) [inline pkg/clients/grpc/connect.go:67]: backendinfo and attested tls are only for agent client(CLI). This needs to bbe  distinguished. Use interfaces for configuration purposes. clienttls is common for all clients and should be part of base config

Comment (SammyOina) [inline pkg/clients/grpc/connect.go:67]: backendinfo and attested tls are only for agent client(CLI). This needs to bbe  distinguished. Use interfaces for configuration purposes. clienttls is common for all clients and should be part of base config

**Review by SammyOina — CHANGES_REQUESTED** [review]

File: pkg/clients/grpc/connect.go [hunk 8]
```diff
@@ -49,14 +49,46 @@ var (
...
+}
+
+type AgentClientConfig struct {
+	ClientConfig
+	AttestationPolicy string `env:"ATTESTATION_POLICY" envDefault:""`
+}
+
+type ManagerClientConfig struct {
+	ClientConfig
+	AttestedTLS bool `env:"ATTESTED_TLS"       envDefault:"false"`
```
Comment (SammyOina) [inline pkg/clients/grpc/connect.go:79]: manager cannot have attested TLS

File: test/computations/main.go [hunk 9]
```diff
@@ -130,7 +130,11 @@ func main() {
 		reflection.Register(srv)
 		manager.RegisterManagerServiceServer(srv, managergrpc.NewServer(incomingChan, &svc{logger: logger}))
 	}
-	grpcServerConfig := server.Config{Port: defaultPort}
+	grpcServerConfig := server.ServerConfig{
+		BaseConfig: server.BaseConfig{
+			URL: fmt.Sprintf("localhost:%s", defaultPort),
+		},
+	}
```
Comment (SammyOina) [inline test/computations/main.go:137]: env only has host and port, so set url in code after parsing config

File: pkg/clients/grpc/connect.go [hunk 10]
```diff
@@ -49,14 +49,46 @@ var (
...
+	ClientTLS    bool          `env:"CLIENT_TLS"    envDefault:"false"`
+}
+
+type ClientConfig struct {
+	BaseConfig
+}
+
+type AgentClientConfig struct {
+	ClientConfig
+	AttestationPolicy string `env:"ATTESTATION_POLICY" envDefault:""`
```
Comment (SammyOina) [inline pkg/clients/grpc/connect.go:74]: attested tls bool?

File: pkg/clients/grpc/connect.go [hunk 11]
```diff
@@ -49,14 +49,46 @@ var (
...
+type BaseConfig struct {
+	URL          string        `env:"URL"             envDefault:"localhost:7001"`
+	Timeout      time.Duration `env:"TIMEOUT"         envDefault:"60s"`
+	ClientCert   string        `env:"CLIENT_CERT"     envDefault:""`
+	ClientKey    string        `env:"CLIENT_KEY"      envDefault:""`
+	ServerCAFile string        `env:"SERVER_CA_CERTS" envDefault:""`
+	CertFile     string        `env:"SERVER_CERT"        envDefault:""`
+	KeyFile      string        `env:"SERVER_KEY"         envDefault:""`
+	ClientCAFile string        `env:"CLIENT_CA_CERTS"    envDefault:""`
+	ClientTLS    bool          `env:"CLIENT_TLS"    envDefault:"false"`
```
Comment (SammyOina) [inline pkg/clients/grpc/connect.go:65]: why was this added

File: pkg/clients/grpc/connect.go [hunk 12]
```diff
@@ -49,14 +49,46 @@ var (
...
+}
+
+type BaseConfig struct {
+	URL          string        `env:"URL"             envDefault:"localhost:7001"`
+	Timeout      time.Duration `env:"TIMEOUT"         envDefault:"60s"`
+	ClientCert   string        `env:"CLIENT_CERT"     envDefault:""`
+	ClientKey    string        `env:"CLIENT_KEY"      envDefault:""`
+	ServerCAFile string        `env:"SERVER_CA_CERTS" envDefault:""`
+	CertFile     string        `env:"SERVER_CERT"        envDefault:""`
+	KeyFile      string        `env:"SERVER_KEY"         envDefault:""`
```
Comment (SammyOina) [inline pkg/clients/grpc/connect.go:63]: server certs are not used on client

File: pkg/clients/grpc/connect.go [hunk 13]
```diff
@@ -119,27 +151,31 @@ func (c *client) Connection() *grpc.ClientConn {
...
 	tc := insecure.NewCredentials()
 
-	if cfg.AttestedTLS {
-		err := ReadAttestationPolicy(cfg.AttestationPolicy, &quoteprovider.AttConfigurationSEVSNP)
-		if err != nil {
-			return nil, secure, errors.Wrap(fmt.Errorf("failed to read Attestation Policy"), err)
-		}
+	switch s := cfg.(type) {
+	case AgentClientConfig:
+		if s.ClientTLS {
```
Comment (SammyOina) [inline pkg/clients/grpc/connect.go:163]: this code is breaking existing functionality

File: pkg/clients/grpc/connect.go [hunk 14]
```diff
@@ -119,27 +147,31 @@ func (c *client) Connection() *grpc.ClientConn {
...
-			InsecureSkipVerify:    true,
-			VerifyPeerCertificate: verifyPeerCertificateATLS,
+			tlsConfig := &tls.Config{
+				InsecureSkipVerify:    true,
+				VerifyPeerCertificate: verifyPeerCertificateATLS,
+			}
+			tc = credentials.NewTLS(tlsConfig)
+			opts = append(opts, grpc.WithContextDialer(CustomDialer))
+			secure = withaTLS
 		}
```
Comment (SammyOina) [inline pkg/clients/grpc/connect.go:172]: will fail if agentconfig is any other config other than atls

File: pkg/clients/grpc/connect.go [hunk 15]
```diff
@@ -49,14 +49,42 @@ var (
...
+	AttestedTLS       bool   `env:"ATTESTED_TLS"       envDefault:"false"`
+}
+
+type ManagerClientConfig struct {
+	ClientConfig
+}
+
+func (s ClientConfig) GetBaseConfig() BaseConfig {
+	return s.BaseConfig
+}
```
Comment (SammyOina) [inline pkg/clients/grpc/connect.go:80]: why do you need both client and base config, clients are only two CLI(Agent) and manager

**Review by SammyOina — CHANGES_REQUESTED** [review]

File: pkg/clients/grpc/connect.go [hunk 16]
```diff
@@ -119,67 +143,92 @@ func (c *client) Connection() *grpc.ClientConn {
...
 				}
-				tlsConfig.Certificates = []tls.Certificate{certificate}
-				secure = withmTLS
+				tc = credentials.NewTLS(tlsConfig)
+				secure = sec
+			}
+		}
+	default:
+		conf := cfg.GetBaseConfig()
+		if conf.ServerCAFile != "" {
```
Comment (SammyOina) [inline pkg/clients/grpc/connect.go:180]: repeated in loadTLSConfig

File: pkg/clients/grpc/connect.go [hunk 17]
```diff
@@ -119,67 +143,92 @@ func (c *client) Connection() *grpc.ClientConn {
...
-				certificate, err := tls.LoadX509KeyPair(cfg.ClientCert, cfg.ClientKey)
+			tlsConfig := &tls.Config{
+				InsecureSkipVerify:    true,
+				VerifyPeerCertificate: verifyPeerCertificateATLS,
+			}
+			tc = credentials.NewTLS(tlsConfig)
+			opts = append(opts, grpc.WithContextDialer(CustomDialer))
+			secure = withaTLS
+		} else {
+			if s.ServerCAFile != "" {
```
Comment (SammyOina) [inline pkg/clients/grpc/connect.go:169]: same here

**Review by SammyOina — CHANGES_REQUESTED** [review]

File: internal/server/grpc/grpc_test.go [hunk 18]
```diff
@@ -52,7 +56,7 @@ func TestNew(t *testing.T) {
 	assert.IsType(t, &Server{}, srv)
 }
 
-func TestServerStartWithTLSFile(t *testing.T) {
+func TestServerStartWithTLS(t *testing.T) {
```
Comment (SammyOina) [inline internal/server/grpc/grpc_test.go:59]: Do not rename the test name here

File: internal/server/grpc/grpc_test.go [hunk 19]
```diff
@@ -80,11 +84,62 @@ func TestServerStartWithTLSFile(t *testing.T) {
...
+	cancel()
+
+	time.Sleep(200 * time.Millisecond)
+
+	logContent := logBuffer.String()
+	fmt.Println(logContent)
+	assert.Contains(t, logContent, "TestServer service gRPC server listening at localhost:0 with TLS")
+}
+
+func TestServerStartWithMTLS(t *testing.T) {
```
Comment (SammyOina) [inline internal/server/grpc/grpc_test.go:127]: Already in another test cas

**Review by SammyOina — CHANGES_REQUESTED** [review]

**Review by drasko — APPROVED** [review]
LGTM

**Pull request merged**